File: assignment1_20201466/assignment1_20201466.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>

/* 전역 변수 선언 */
int charClass;
char lexeme[100];
char input_str[2048];
char nextChar;
int lexLen;
int token;
int nextToken;
int input_str_index;
int SYNTAX_ERROR;
int bexpr_error;
int expr_index;
int notdecimal;
int dec_dect;
char demical[100];
int demical_index;
int term_data;
int aexpr_data;

/* 계산 후 저장할 배열 */
int print_index;
int print_final[100];

/* 계산 결과 저장할 배열 */
int print_x[100];           // 1번
int print_y[100];           // 2번
int print_z[100];           // 3번

bool print_x_bool[100];     // 4번
bool print_y_bool[100];     // 5번
bool print_z_bool[100];     // 6번

/* 임시 저장공간 */
int var_x;
int var_y;
int var_z;

/* 변수(x, y, z)와 연산자 확인 변수 */
int var_detect;
int cmp_detect;

bool moving_bool;
int moving_dec;

int bool_dec;

int bool_dec_x;
int bool_dec_y;
int bool_dec_z;

/* 함수 */
/* 문자열 추출 함수 */
int lookup_op_1(char ch);
int lookup_op_2(char ch);
int lookup_var(char ch);
void addChar();
void getChar();
void getNonBlank();
int lex();

/* 파싱 함수 */
void program();
void statement();
void expr();
void bexpr();
void relop();
void aexpr();
void term();
void factor();
void number();
void dec();
void var();

/* 에러 함수 */
void error();



/* Character */
#define LETTER 1
#define DIGIT 2
#define UNKNOWN 99

/*
 * 토큰
 * ' +, -, *, /, =, !, <, >, ; '
 */
#define NULL_A 0
#define INT_LIT 10
#define IDENT 11
#define OPERATION 12

/* 연산자 */
#define ASSIGN_OP 20
#define ADD_OP 21
#define SUB_OP 22
#define MULT_OP 23
#define DIV_OP 24
#define LESS_THAN 25
#define LESS 26
#define MORE_THAN 27
#define MORE 28
#define EQUAL 29
#define NOT_EQUAL 30
#define NOT 31
#define SEMI_COLON 32

/* 변수 */
#define VAR_X 33
#define VAR_Y 34
#define VAR_Z 35


int main() {
    while (1) {
        SYNTAX_ERROR = 0;
        bexpr_error = 0;
        input_str_index = 0;
        var_x=0; var_y=0; var_z=0;
        print_index = 0;
        bool_dec = 1;
        bool_dec_x = 1; bool_dec_y = 1; bool_dec_z = 1;
        memset(input_str, '\0', sizeof(input_str));
        printf(">> ");
        fgets(input_str, 2048, stdin);

        if (strcmp(input_str, "terminate\n") == 0) {
            exit(0);
        }

        getChar();

        /*
         * RD 파싱 시작
         */
        program();

        /*
         *  출력
         */
        printf(">> ");
        if (SYNTAX_ERROR == 0) {
            int length_print = print_index;
            for (int i=0; i<print_index; i++) {
                switch (print_final[i]) {
                    // INT 형 x 출력
                    case 1 :
                        printf("%d ", print_x[i]);
                        break;

                    // int 형 y 출력
                    case 2 :
                        printf("%d ", print_y[i]);
                        break;

                    // int 형 z 출력
                    case 3 :
                        printf("%d ", print_z[i]);
                        break;

                    // bool 형 x 출력
                    case 4 :
                        if (print_x_bool[i] == 0) {
                            printf("FALSE ");
                        }
                        else {
                            printf("TRUE ");
                        }
                        break;

                    // bool 형 y 출력
                    case 5 :
                        if (print_y_bool[i] == 0) {
                            printf("FALSE ");
                        }
                        else {
                            printf("TRUE ");
                        }
                        break;

                    // bool 형 z 출력
                    case 6 :
                        if (print_z_bool[i] == 0) {
                            printf("FALSE ");
                        }
                        else {
                            printf("TRUE ");
                        }
                        break;

                    default :
                        break;
                }
            }
            printf("\n");
        }
        else {
            printf("syntax error!!\n");
        }
    }
    return 0;
}

/*
 *  연산자가 처음 올 때 어떤 연산자인지 확인하여 정보를 저장하는 함수
 */
int lookup_op_1(char ch) {
    switch (ch) {
        case '+' :
            addChar();
            nextToken = ADD_OP;
            break;

        case '-' :
            addChar();
            nextToken = SUB_OP;
            break;

        case '*' :
            addChar();
            nextToken = MULT_OP;
            break;

        case '/' :
            addChar();
            nextToken = DIV_OP;
            break;

        case '=' :
            addChar();
            nextToken = ASSIGN_OP;
            break;

        case '!' :
            addChar();
            nextToken = NOT;
            break;

        case '<' :
            addChar();
            nextToken = LESS_THAN;
            break;

        case '>' :
            addChar();
            nextToken = MORE_THAN;
            break;

        case ';' :
            addChar();
            nextToken = SEMI_COLON;
            break;

        default :
            addChar();
            nextToken = OPERATION;
            break;
    }
    return nextToken;
}

/*
 *   연산자가 2개 이상 연속으로 올 때 자신이 쓰는
 */
int lookup_op_2(char ch) {
    switch (ch) {
        case '=' :
            // '=='
            if (nextToken == ASSIGN_OP) {
                addChar();
                nextToken = EQUAL;
            }

            // '!='
            else if (nextToken == NOT) {
                addChar();
                nextToken = NOT_EQUAL;
            }

            // '<='
            else if (nextToken == LESS_THAN) {
                addChar();
                nextToken = LESS;
            }

            // '>='
            else if (nextToken == MORE_THAN) {
                addChar();
                nextToken = MORE;
            }

            // 기타 등등
            else {
                addChar();
                nextToken = OPERATION;
            }
            break;

            // 기타 등등
        default :
            addChar();
            nextToken = OPERATION;
            break;
    }
    return nextToken;
}

/*
 * <var> 결과 값을 확인 하려는 함수
 */
int lookup_var(char ch) {
    switch (ch) {
        case 'x' :
            addChar();
            nextToken = VAR_X;
            break;

        case 'y' :
            addChar();
            nextToken = VAR_Y;
            break;

        case 'z' :
            addChar();
            nextToken = VAR_Z;
            break;

        default :
            addChar();
            nextToken = IDENT;
            break;
    }

    return nextToken;
}

/*
 *  분리한 토큰을 배열에 저장하는 함수
 */
void addChar() {
    if (lexLen <= 98) {
        lexeme[lexLen++] = nextChar;
        lexeme[lexLen] = 0;
    }
    else {
        printf("error");
    }
}

/*
 *   판단할 문자가 문자인지, 숫자인지, 연산자인지 판단하는 함수
 */
void getChar() {
    if ((nextChar = input_str[input_str_index++]) != '\0') {
        if (isalpha(nextChar)) {
            charClass = LETTER;
        }
        else if (isdigit(nextChar)) {
            charClass = DIGIT;
        }
        else { charClass = UNKNOWN; }
    }
    else {
        charClass = NULL_A;
    }
}

/*
 *   공백 키를 건너 띄어주는 함수
 */
void getNonBlank() {
    while (isspace(nextChar)) {
        getChar();
    }
}

/*
 *   분리한 토큰의 정보를 저장해주는 함수
 */
int lex() {
    lexLen = 0;
    getNonBlank();

    switch (charClass) {
        case LETTER :
            lookup_var(nextChar);
            getChar();

            while (!isspace(nextChar)) {
                addChar();
                getChar();
                nextToken = IDENT;
            }

            break;

        case DIGIT :
            addChar();
            getChar();

            nextToken = INT_LIT;
            break;

        case UNKNOWN :
            lookup_op_1(nextChar);
            getChar();

            while (!isspace(nextChar)) {
                lookup_op_2(nextChar);
                getChar();
            }

            break;

        case NULL_A :
            nextToken = NULL_A;
            lexeme[0] = 'N';
            lexeme[1] = 'U';
            lexeme[2] = 'L';
            lexeme[3] = 'L';
            lexeme[4] = 0;
            break;
    }
//    printf("next token : %d, next lexeme : %s\n", nextToken, lexeme);
    return nextToken;
}

/*
 *  <program> -> {<statement>}
 */
void program() {
    lex();

    while (nextToken != NULL_A && SYNTAX_ERROR == 0) {
        statement();
    }
}

/*
 *  <statement> -> <var> = <expr> ; | print <var> ;
 */
void statement() {
    if (SYNTAX_ERROR == 0) {
        // <var> = <expr> ;
        if (strcmp(lexeme, "print") != 0) {
            var();

            if (SYNTAX_ERROR == 0) {
                if (nextToken != ASSIGN_OP) {
                    error();
                }
                else {
                    expr_index = input_str_index;
                    lex();
                    expr();
                }
            }

            if (SYNTAX_ERROR == 0) {
                if (nextToken != SEMI_COLON) {
                    error();
                }
                else {
                    lex();
                }
            }

            // <expr> 의 결과 값을 저장
            if (SYNTAX_ERROR == 0) {
                // 결과 값이 bool 형일 때
                if (bool_dec == 0) {
                    switch (var_detect) {
                        case VAR_X :
                            bool_dec_x = 0;
                            var_x = moving_bool;
                            break;

                        case VAR_Y :
                            bool_dec_y = 0;
                            var_y = moving_bool;
                            break;

                        case VAR_Z :
                            bool_dec_z = 0;
                            var_z = moving_bool;
                            break;

                        default :
                            break;
                    }
                }
                // 결과 값이 int 형일 때
                else if (bool_dec == 1){
                    switch (var_detect) {
                        case VAR_X :
                            bool_dec_x = 1;
                            var_x = moving_dec;
                            break;

                        case VAR_Y :
                            bool_dec_y = 1;
                            var_y = moving_dec;
                            break;

                        case VAR_Z :
                            bool_dec_z = 1;
                            var_z = moving_dec;
                            break;

                        default :
                            break;
                    }
                }
            }
        }
        // print <var> ;
        else {
            lex();
            var();

            if (SYNTAX_ERROR == 0) {
                if (nextToken != SEMI_COLON) {
                    error();
                }
                else {
                    lex();
                }
            }

            if (SYNTAX_ERROR == 0) {
                // x, y, z 구별
                switch (var_detect) {
                    case VAR_X :
                        // x 가 bool 형 일 때
                        if (bool_dec_x == 0) {
                            print_x_bool[print_index] = var_x;
                            print_final[print_index++] = 4;
                            print_final[print_index] = 0;
                        }
                        // x 가 int 형일 때
                        else if (bool_dec_x == 1) {
                            print_x[print_index] = var_x;
                            print_final[print_index++] = 1;
                            print_final[print_index] = 0;
                        }
                        break;

                    case VAR_Y :
                        // y 가 bool형 일 때
                        if (bool_dec_y == 0) {
                            print_y_bool[print_index] = var_y;
                            print_final[print_index++] = 5;
                            print_final[print_index] = 0;
                        }
                        // y 가 int 형 일 때
                        else if (bool_dec_y == 1) {
                            print_y[print_index] = var_y;
                            print_final[print_index++] = 2;
                            print_final[print_index] = 0;
                        }
                        break;

                    case VAR_Z :
                        // z 가 bool형 일 때
                        if (bool_dec_z == 0) {
                            print_z_bool[print_index] = var_z;
                            print_final[print_index++] = 6;
                            print_final[print_index] = 0;
                        }
                        // z 가 int 형일 때
                        else if (bool_dec_z == 1) {
                            print_z[print_index] = var_z;
                            print_final[print_index++] = 3;
                            print_final[print_index] = 0;
                        }
                        break;

                    default :
                        break;
                }
            }
        }
    }
}

/*
 *  <expr> -> <bexpr> | <aexpr>
 */
void expr() {
    if (SYNTAX_ERROR == 0) {
        bexpr_error = 0;

        // 먼저 실행
        bexpr();

        // <bexpr> 이 오류가 날 시 <aexpr> 실행
        if (bexpr_error == -1) {
            SYNTAX_ERROR = 0;
            bexpr_error = 0;
            input_str_index = expr_index;
            getChar();
            lex();
            aexpr();
        }
    }
}

/*
 *  <bexpr> -> <number> <relop> <number>
 */
void bexpr() {
    if (SYNTAX_ERROR == 0) {
        number();
        int front = atoi(demical);
        demical[0] = 0;

        if (SYNTAX_ERROR == 0) {
            relop();
        }

        if (SYNTAX_ERROR == 0) {
            number();
        }

        // 숫자를 변환하고 연산자의 판별을 통해 계산 (bool 형)
        if (SYNTAX_ERROR == 0) {
            int back = atoi(demical);
            demical[0] = 0;

            int bool_bexpr;

            switch (cmp_detect) {
                case EQUAL :
                    bool_bexpr = (front == back);
                    break;

                case NOT_EQUAL :
                    bool_bexpr = (front != back);
                    break;

                case MORE :
                    bool_bexpr = (front >= back);
                    break;

                case MORE_THAN :
                    bool_bexpr = (front > back);
                    break;

                case LESS :
                    bool_bexpr = (front <= back);
                    break;

                case LESS_THAN :
                    bool_bexpr = (front < back);
                    break;

                default :
                    break;
            }
            // 전역 변수에 저장
            moving_bool = bool_bexpr;
            bool_dec = 0;
        }
   }
}

/*
 *  <relop> -> == | != | < | > | <= | >=
 */
void relop() {
    if (SYNTAX_ERROR == 0) {
        switch (nextToken) {
            case EQUAL :
                cmp_detect = EQUAL;
                break;

            case NOT_EQUAL :
                cmp_detect = NOT_EQUAL;
                break;

            case MORE :
                cmp_detect = MORE;
                break;

            case MORE_THAN :
                cmp_detect = MORE_THAN;
                break;

            case LESS :
                cmp_detect = LESS;
                break;

            case LESS_THAN :
                cmp_detect = LESS_THAN;
                break;

            default :
                error();
                bexpr_error = -1;
                // lex();
                break;
        }
        if (SYNTAX_ERROR == 0) {
            lex();
        }
    }
}

/*
 *  <aexpr> -> <term> {(+ | -) <term>}
 */
void aexpr() {
    int front=0, back=0, operand=0;
    if (SYNTAX_ERROR == 0) {
        term();
        front = term_data;

        // 에러가 나지 않았거나 +, - 가 오면 반복
        while ((nextToken == ADD_OP || nextToken == SUB_OP) && SYNTAX_ERROR==0) {
            if (nextToken == ADD_OP) {
                operand = ADD_OP;
            }
            else if (nextToken == SUB_OP) {
                operand = SUB_OP;
            }
            lex();
            term();

            // 계산
            back = term_data;
            if (operand == ADD_OP) {
                front = front + back;
            }
            else if (operand == SUB_OP) {
                front = front - back;
            }
        }
    }
    if (SYNTAX_ERROR == 0) {
        bool_dec = 1;
        moving_dec = front;
    }
}

/*
 *  <term> -> <factor> {(* | /) <factor>
 */
void term() {
    int front=0, back=0, operand;
    if (SYNTAX_ERROR == 0) {
        factor();
        front = atoi(demical);

        // 에러가 나지 않았거나 *, / 가 오면 반복
        while ((nextToken == MULT_OP || nextToken == DIV_OP) && SYNTAX_ERROR==0) {
            if (nextToken == MULT_OP) {
                operand = MULT_OP;
            }
            else if (nextToken == DIV_OP) {
                operand = DIV_OP;
            }
            lex();
            factor();

            // 계산
            back = atoi(demical);
            if (operand == MULT_OP) {
                front = front * back;
            }
            else if (operand == DIV_OP) {
                front = front / back;
            }
        }
    }
    if (SYNTAX_ERROR == 0) {
        term_data = front;
    }
}

/*
 *  <factor> -> <number>
 */
void factor() {
    if (SYNTAX_ERROR == 0) {
        number();
    }
}

/*
 *  <number> -> <dec>{<dec>}
 */
void number() {
    dec_dect = 0;
    memset(demical, '\0', sizeof(demical));
    if (SYNTAX_ERROR == 0) {
        dec();
        dec_dect++;

        // 정수이거나 에러가 나지 않았을 때
        while (notdecimal==0 && SYNTAX_ERROR == 0) {
            dec();
        }
        notdecimal = 0;
        demical_index = 0;
    }
}

/*
 * <dec> -> 0|1|2|3|4|5|6|7|8|9
 */
void dec() {
    if (SYNTAX_ERROR == 0) {
        if (nextToken == INT_LIT) {
            demical[demical_index++] = lexeme[0];
            lex();
        }
        // 첫번째 <dec> 에 정수가 아니면 에러
        else if (dec_dect == 0) {
            error();
            bexpr_error = -1;
        }
        // 이후의 <dec>은 지금 판별하진 않음
        else {
            notdecimal = -1;
        }
    }
}

/*
 *  <var> -> x | y | z
 */
void var() {
    if (SYNTAX_ERROR == 0) {
        if (nextToken == VAR_X || nextToken == VAR_Y || nextToken == VAR_Z) {
            if (nextToken == VAR_X) {
                var_detect = VAR_X;
            }
            else if (nextToken == VAR_Y) {
                var_detect = VAR_Y;
            }
            else if (nextToken == VAR_Z) {
                var_detect = VAR_Z;
            }
            lex();
        }
        else {
            error();
        }
    }
}

void error() {
    SYNTAX_ERROR = -1;
}

File: assignment1_20201466/assignment1_20201466.py
# 문자 분류 상수
LETTER = 1
DIGIT = 2
UNKNOWN = 99

# 토큰 상수
NULL = 0
INT_LIT = 10
IDENT = 11
OPERATION = 12

# 연산자 상수
ASSIGN_OP = 20
ADD_OP = 21
SUB_OP = 22
MULT_OP = 23
DIV_OP = 24
LESS_THAN = 25
LESS = 26
MORE_THAN = 27
MORE = 28
EQUAL = 29
NOT_EQUAL = 30
NOT = 31
SEMI_COLON = 32

# 변수 상수
VAR_X = 33
VAR_Y = 34
VAR_Z = 35

# 전역 변수
charClass = 0
lexeme = ""
input_str = ""
nextChar = ""
lexLen = 0
token = 0
nextToken = 0
input_str_index = 0
SYNTAX_ERROR = 0
bexpr_error = 0
expr_index = 0
notdecimal = 0
dec_dect = 0
demical = ""
demical_index = 0
term_data = 0
aexpr_data = 0

# 계산 결과 저장 배열
# print_index = 0
print_final = []
print_x = []
print_y = []
print_z = []

# 임시 저장 변수
var_x = 0
var_y = 0
var_z = 0

# 변수 및 연산자 확인 변수
var_detect = 0
cmp_detect = 0

moving_bool = False
moving_dec = 0

bool_dec = 1
bool_dec_x = 1
bool_dec_y = 1
bool_dec_z = 1

# 연산자 확인 첫번째
def lookup_op_1(ch):
    global nextToken, ADD_OP, SUB_OP, MULT_OP, DIV_OP, ASSIGN_OP, NOT, LESS_THAN, MORE_THAN, SEMI_COLON, OPERATION
    if ch == '+':
        addChar()
        nextToken = ADD_OP
    elif ch == '-':
        addChar()
        nextToken = SUB_OP
    elif ch == '*':
        addChar()
        nextToken = MULT_OP
    elif ch == '/':
        addChar()
        nextToken = DIV_OP
    elif ch == '=':
        addChar()
        nextToken = ASSIGN_OP
    elif ch == '!':
        addChar()
        nextToken = NOT
    elif ch == '<':
        addChar()
        nextToken = LESS_THAN
    elif ch == '>':
        addChar()
        nextToken = MORE_THAN
    elif ch == ';':
        addChar()
        nextToken = SEMI_COLON
    else:
        addChar()
        nextToken = OPERATION
    return nextToken


# 연산자 확인 두번째 이후
def lookup_op_2(ch):
    global nextToken, ASSIGN_OP, EQUAL, NOT, NOT_EQUAL, LESS, LESS_THAN, MORE, MORE_THAN, OPERATION
    if ch == '=':
        if nextToken == ASSIGN_OP:
            addChar()
            nextToken = EQUAL
        elif nextToken == NOT:
            addChar()
            nextToken = NOT_EQUAL
        elif nextToken == LESS_THAN:
            addChar()
            nextToken = LESS
        elif nextToken == MORE_THAN:
            addChar()
            nextToken = MORE
        else:
            addChar()
            nextToken = OPERATION
    else:
        addChar()
        nextToken = OPERATION
    return nextToken

# x, y, z 구분
def lookup_var(ch):
    global nextToken, VAR_X, VAR_Y, VAR_Z

    if ch == 'x':
        addChar()
        nextToken = VAR_X
    elif ch == 'y':
        addChar()
        nextToken = VAR_Y
    elif ch == 'z':
        addChar()
        nextToken = VAR_Z
    else:
        addChar()
        nextToken = IDENT

    return nextToken

# 임시버퍼에 토큰 저장
def addChar():
    global lexLen, lexeme, nextChar
    lexeme = lexeme + nextChar

# 토큰이 될 때까지 문자 읽기
def getChar():
    global input_str_index, nextChar, charClass, input_str, LETTER, DIGIT, UNKNOWN, NULL

    if input_str_index < len(input_str):
        nextChar = input_str[input_str_index]
        input_str_index += 1
        if nextChar.isalpha():
            charClass = LETTER
        elif nextChar.isdigit():
            charClass = DIGIT
        elif nextChar == '\0' :
            charClass = NULL
        else:
            charClass = UNKNOWN
    else:
        charClass = NULL

# 공백 무시
def getNonBlank():
    while nextChar.isspace():
        getChar()

# 토큰의 정보 저장
def lex():
    global lexLen, nextToken, lexeme, nextChar, charClass, LETTER, DIGIT, UNKNOWN, NULL
    lexeme = ""
    lexLen = 0
    getNonBlank()

    if charClass == LETTER:

        lookup_var(nextChar)
        getChar()

        while nextChar != '\0' and not nextChar.isspace():
            addChar()
            getChar()
            nextToken = IDENT

    elif charClass == DIGIT:
        addChar()
        getChar()

        nextToken = INT_LIT

    elif charClass == UNKNOWN:
        lookup_op_1(nextChar)
        getChar()

        while nextChar != '\0' and not nextChar.isspace():
            lookup_op_2(nextChar)
            getChar()

    elif charClass == NULL:
        nextToken = NULL
        lexeme = 'NULL'

    # print(f"next token: {nextToken}, next lexeme: {lexeme}")
    return nextToken

# <program> -> {<statement>}
def program():
    lex()

    while nextToken != NULL and SYNTAX_ERROR == 0:
        statement()

# <statement> -> <var> = <expr> ; | print <var> ;
def statement():
    global SYNTAX_ERROR, lexeme, nextToken, input_str_index, expr_index, bool_dec, bool_dec_x, bool_dec_y, bool_dec_z, var_detect, var_x, var_y, var_z, VAR_X, VAR_Y, VAR_Z
    global print_x, print_y, print_z, print_x_bool, print_y_bool, print_z_bool, print_index, print_final, moving_bool, moving_dec
    if SYNTAX_ERROR == 0:
        # <var> = <expr> ;
        if lexeme != "print":
            var()

            if SYNTAX_ERROR == 0:
                if nextToken != ASSIGN_OP:
                    error()
                else:
                    expr_index = input_str_index
                    lex()
                    expr()

            if SYNTAX_ERROR == 0:
                if nextToken != SEMI_COLON:
                    error()
                else:
                    lex()

            # <expr> 의 결과 값을 저장
            if SYNTAX_ERROR == 0:
                # 결과 값이 bool 형일 때
                if bool_dec == 0:
                    if var_detect == VAR_X:
                        bool_dec_x = 0
                        var_x = moving_bool
                    elif var_detect == VAR_Y:
                        bool_dec_y = 0
                        var_y = moving_bool
                    elif var_detect == VAR_Z:
                        bool_dec_z = 0
                        var_z = moving_bool
                # 결과 값이 int 형일 때
                elif bool_dec == 1:
                    if var_detect == VAR_X:
                        bool_dec_x = 1
                        var_x = moving_dec
                    elif var_detect == VAR_Y:
                        bool_dec_y = 1
                        var_y = moving_dec
                    elif var_detect == VAR_Z:
                        bool_dec_z = 1
                        var_z = moving_dec
        # print <var> ;
        else:
            lex()
            var()

            if SYNTAX_ERROR == 0:
                if nextToken != SEMI_COLON:
                    error()
                else:
                    lex()

            if SYNTAX_ERROR == 0:
                # x, y, z 구별
                if var_detect == VAR_X:
                    print_x.append(var_x)
                    print_final.append(1)

                elif var_detect == VAR_Y:
                    print_y.append(var_y)
                    print_final.append(2)

                elif var_detect == VAR_Z :
                    print_z.append(var_z)
                    print_final.append(3)

# <expr> -> <bexpr> | <aexpr>
def expr():
    global SYNTAX_ERROR, bexpr_error, expr_index, input_str_index

    if SYNTAX_ERROR == 0:
        bexpr_error = 0

        # 먼저 실행
        bexpr()

        # <bexpr> 이 오류가 날 시 <aexpr> 실행
        if bexpr_error == -1:
            SYNTAX_ERROR = 0
            bexpr_error = 0
            input_str_index = expr_index
            getChar()
            lex()
            aexpr()

# <bexpr> -> <number> <relop> <number>
def bexpr():
    global SYNTAX_ERROR, demical, cmp_detect, moving_bool, bool_dec

    if SYNTAX_ERROR == 0:
        number()
        if (len(demical)==0)  :
            error()
        else :
            front = int(demical)

        if SYNTAX_ERROR == 0:
            relop()

        if SYNTAX_ERROR == 0:
            number()

        # 숫자를 변환하고 연산자의 판별을 통해 계산 (bool 형)
        if SYNTAX_ERROR == 0:
            if (len(demical) == 0):
                error()
            else :
                back = int(demical)

            bool_bexpr = 0

            if cmp_detect == EQUAL:
                bool_bexpr = (front == back)
            elif cmp_detect == NOT_EQUAL:
                bool_bexpr = (front != back)
            elif cmp_detect == MORE:
                bool_bexpr = (front >= back)
            elif cmp_detect == MORE_THAN:
                bool_bexpr = (front > back)
            elif cmp_detect == LESS:
                bool_bexpr = (front <= back)
            elif cmp_detect == LESS_THAN:
                bool_bexpr = (front < back)

            # 전역 변수에 저장
            moving_bool = bool_bexpr
            bool_dec = 0

# <relop> -> == | != | < | > | <= | >=
def relop():
    global SYNTAX_ERROR, nextToken, cmp_detect, bexpr_error

    if SYNTAX_ERROR == 0:
        if nextToken == EQUAL:
            cmp_detect = EQUAL
        elif nextToken == NOT_EQUAL:
            cmp_detect = NOT_EQUAL
        elif nextToken == MORE:
            cmp_detect = MORE
        elif nextToken == MORE_THAN:
            cmp_detect = MORE_THAN
        elif nextToken == LESS:
            cmp_detect = LESS
        elif nextToken == LESS_THAN:
            cmp_detect = LESS_THAN
        else:
            error()
            bexpr_error = -1

        if SYNTAX_ERROR == 0:
            lex()

# <aexpr> -> <term> {(+ | -) <term>}
def aexpr():
    global SYNTAX_ERROR, nextToken, term_data, bool_dec, moving_dec, ADD_OP, SUB_OP

    front = 0
    back = 0
    operand = 0

    if SYNTAX_ERROR == 0:
        term()
        front = term_data

        # 에러가 나지 않았거나 +, - 가 오면 반복
        while (nextToken == ADD_OP or nextToken == SUB_OP) and SYNTAX_ERROR == 0:
            if nextToken == ADD_OP:
                operand = ADD_OP
            elif nextToken == SUB_OP:
                operand = SUB_OP
            lex()
            term()

            # 계산
            back = term_data
            if operand == ADD_OP:
                front = front + back
            elif operand == SUB_OP:
                front = front - back

    if SYNTAX_ERROR == 0:
        bool_dec = 1
        moving_dec = front

# <term> -> <factor> {(* | /) <factor>
def term():
    global SYNTAX_ERROR, nextToken, demical, term_data, MULT_OP, DIV_OP

    front = 0
    back = 0
    operand = 0

    if SYNTAX_ERROR == 0:
        factor()
        if (len(demical)==0)  :
            error()
        else :
            front = int(demical)

        # 에러가 나지 않았거나 *, / 가 오면 반복
        while (nextToken == MULT_OP or nextToken == DIV_OP) and SYNTAX_ERROR == 0:
            if nextToken == MULT_OP:
                operand = MULT_OP
            elif nextToken == DIV_OP:
                operand = DIV_OP
            lex()
            factor()

            # 계산
            if (len(demical) == 0):
                error()
            else:
                back = int(demical)
            if operand == MULT_OP:
                front = front * back
            elif operand == DIV_OP:
                front = front // back

    if SYNTAX_ERROR == 0:
        term_data = front

# <factor> -> <number>
def factor():
    global SYNTAX_ERROR

    if SYNTAX_ERROR == 0:
        number()

# <number> -> <dec>{<dec>}
def number():
    global dec_dect, demical, SYNTAX_ERROR, notdecimal, demical_index

    dec_dect = 0
    demical = ""

    if SYNTAX_ERROR == 0:
        dec()
        dec_dect += 1

        while notdecimal == 0 and SYNTAX_ERROR == 0:
            dec()

        notdecimal = 0
        demical_index = 0

# <dec> -> 0|1|2|3|4|5|6|7|8|9
def dec():
    global SYNTAX_ERROR, nextToken, lexeme, demical, demical_index, dec_dect, bexpr_error, notdecimal

    if SYNTAX_ERROR == 0:
        if nextToken == INT_LIT:
            demical = demical + lexeme[0]
            demical_index += 1
            lex()
        elif dec_dect == 0:
            error()
            bexpr_error = -1
        else:
            notdecimal = -1

# <var> -> x | y | z
def var():
    global SYNTAX_ERROR, nextToken, var_detect, VAR_X, VAR_Y, VAR_Z

    if SYNTAX_ERROR == 0:
        if nextToken in [VAR_X, VAR_Y, VAR_Z]:
            if nextToken == VAR_X:
                var_detect = VAR_X
            elif nextToken == VAR_Y:
                var_detect = VAR_Y
            else:
                var_detect = VAR_Z
            lex()
        else:
            error()


def error():
    global SYNTAX_ERROR
    SYNTAX_ERROR = -1



if __name__ == "__main__":
    while True:
        # 초기화
        SYNTAX_ERROR = 0
        bexpr_error = 0
        input_str_index = 0
        var_x = 0
        var_y = 0
        var_z = 0
        print_index = 0
        bool_dec = 1
        bool_dec_x = 1
        bool_dec_y = 1
        bool_dec_z = 1
        input_str = ""
        print_x = []
        print_y = []
        print_z = []
        print_final = []

        # 사용자 입력 받기
        print(">> ", end="")
        input_str = input()

        if input_str.strip() == "terminate":
            break

        input_str = input_str + '\0'
        getChar()

        # RD 파싱 시작
        program()

        # 출력
        print(">> ", end="")
        if SYNTAX_ERROR == 0:
            print_x_index = 0
            print_y_index = 0
            print_z_index = 0

            for i in range(len(print_final)):
                if print_final[i] == 1:
                    print(print_x[print_x_index], end=" ")
                    print_x_index += 1
                elif print_final[i] == 2:
                    print(print_y[print_y_index], end=" ")
                    print_y_index += 1
                elif print_final[i] == 3:
                    print(print_z[print_z_index], end=" ")
                    print_z_index += 1
            print()
        else:
            print("syntax error!!")
